src/ble/BleLayer.cpp

```diff
@@ -347,25 +340,35 @@ BLE_ERROR BleLayer::Shutdown()
             }
         }
     }
-
     return BLE_NO_ERROR;
 }
 
-BLE_ERROR BleLayer::NewBleConnection(void * appState, const uint16_t connDiscriminator,
-                                     BleConnectionDelegate::OnConnectionCompleteFunct onConnectionComplete,
-                                     BleConnectionDelegate::OnConnectionErrorFunct onConnectionError)
+BLE_ERROR BleLayer::CloseBleConnection(BLE_CONNECTION_OBJECT connObj)
 {
-    BLE_ERROR err = BLE_NO_ERROR;
-
-    VerifyOrExit(mState == kState_Initialized, err = BLE_ERROR_INCORRECT_STATE);
-    VerifyOrExit(mConnectionDelegate != nullptr, err = BLE_ERROR_INCORRECT_STATE);
+    // Close and free all BLE end points.
+    for (size_t i = 0; i < BLE_LAYER_NUM_BLE_ENDPOINTS; i++)
+    {
+        BLEEndPoint * elem = sBLEEndPointPool.Get(i);
 
-    mConnectionDelegate->OnConnectionComplete = onConnectionComplete;
-    mConnectionDelegate->OnConnectionError    = onConnectionError;
-    mConnectionDelegate->NewConnection(this, appState, connDiscriminator);
+        // If end point was initialized, and has not since been freed...
+        if (elem->mBle != nullptr && elem->ConnectionObjectIs(connObj))
+        {
+            // If end point hasn't already been closed...
+            if (elem->mState != BLEEndPoint::kState_Closed)
+            {
+                // Close end point such that callbacks are suppressed and pending transmissions aborted.
+                elem->Abort();
+            }
 
-exit:
-    return err;
+            // If end point was closed, but is still waiting for GATT unsubscribe to complete, free it anyway.
+            // This cancels the unsubscribe timer (plus all the end point's other timers).
+            if (elem->IsUnsubscribePending())
+            {
+                elem->Free();
+            }
```

Comment:
Is it possible to move the call to `Free` somewhere in `Abort` ? It makes it harder to follow the memory management of the `BLEEndPoint` with some calls to `Free` outside of it.

Reply:
This is placed in BleLayer::Shutdown() before, since we are not discussing the actual layout of BLELayer.cpp in this PR, we can create an issue to collect all BLELayer related improvments.

Reply:
Fair enough.